certification/internal/engine/engine_test.go

```diff
@@ -1,10 +1,111 @@
 package engine
 
 import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http/httptest"
+	"net/url"
+
+	"github.com/google/go-containerregistry/pkg/crane"
+	"github.com/google/go-containerregistry/pkg/registry"
+	"github.com/google/go-containerregistry/pkg/v1/random"
 	. "github.com/onsi/ginkgo/v2"
 	. "github.com/onsi/gomega"
+	"github.com/redhat-openshift-ecosystem/openshift-preflight/certification"
 )
 
+var _ = Describe("Execute Checks tests", func() {
+	var src string
+	var engine CraneEngine
+	BeforeEach(func() {
+		// Set up a fake registry.
+		s := httptest.NewServer(registry.New())
+		DeferCleanup(func() {
+			s.Close()
+		})
+		u, err := url.Parse(s.URL)
+		Expect(err).ToNot(HaveOccurred())
+
+		src = fmt.Sprintf("%s/test/crane", u.Host)
+
+		// Expected values.
+		img, err := random.Image(1024, 5)
+		Expect(err).ToNot(HaveOccurred())
+
+		err = crane.Push(img, src)
+		Expect(err).ToNot(HaveOccurred())
+
+		goodCheck := certification.NewGenericCheck(
+			"testcheck",
+			func(context.Context, certification.ImageReference) (bool, error) {
+				return true, nil
+			},
+			certification.Metadata{},
+			certification.HelpText{},
+		)
+
+		errorCheck := certification.NewGenericCheck(
+			"errorCheck",
+			func(context.Context, certification.ImageReference) (bool, error) {
+				return false, errors.New("errorCheck")
+			},
+			certification.Metadata{},
+			certification.HelpText{},
+		)
+
+		failedCheck := certification.NewGenericCheck(
+			"failedCheck",
+			func(context.Context, certification.ImageReference) (bool, error) {
+				return false, nil
+			},
+			certification.Metadata{},
+			certification.HelpText{},
+		)
+
+		engine = CraneEngine{
+			Image: src,
+			Checks: []certification.Check{
+				goodCheck,
+				errorCheck,
+				failedCheck,
+			},
+			IsBundle:  false,
+			IsScratch: false,
+		}
+	})
+	Context("Run the checks", func() {
+		ctx := context.TODO()
+		It("should succeed", func() {
+			err := engine.ExecuteChecks(ctx)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(engine.results.Passed).To(HaveLen(1))
+			Expect(engine.results.Failed).To(HaveLen(1))
+			Expect(engine.results.Errors).To(HaveLen(1))
+			Expect(engine.results.CertificationHash).To(BeEmpty())
+		})
+		Context("it is a bundle", func() {
+			JustBeforeEach(func() {
+				engine.IsBundle = true
+			})
+			It("should succeed and generate a bundle hash", func() {
+				err := engine.ExecuteChecks(context.TODO())
+				Expect(err).ToNot(HaveOccurred())
+				Expect(engine.results.CertificationHash).ToNot(BeEmpty())
+			})
+		})
+		Context("the image is invalid", func() {
+			JustBeforeEach(func() {
+				engine.Image = "does.not/exist/anywhere:ever"
```

Comment:
In relation to the use of `%s/test/crane` earlier, I'm assuming this replaces the image value with a path which the register server cannot serve, hence, why this is a pull failure case?

Reply:
Resolve this comment if I understood it correctly

Reply:
That is correct. It tests the failure path of "not in registry". Probably not a likely scenario, but at least provides coverage of that if err != nil case.